[#44414] Add the Enterprise Edition guard on date alerts settings (1/2)

dombesz opened the pull request:
https://community.openproject.org/work_packages/44414 and some minor refactors.

- [X] Add "date_alert" to the `Authorization::EnterpriseService::GUARDED_ACTIONS`.
- [X] Guard the [UserPreferences::ParamsContract](https://github.com/opf/openproject/blob/dev/app/contracts/user_preferences/params_contract.rb)

This is the part 1 of the whole work package, part 2 will come, once we have the background job, and the date alert settings frontend merged.

dombesz committed:
[#44414] Add the Enterprise Edition guard on date alerts settings

https://community.openproject.org/work_packages/44414

cbliard commented on app/contracts/user_preferences/params_contract.rb:
```diff
@@ -28,16 +28,20 @@
+    include RequiresEnterpriseGuard
+    self.enterprise_action = :date_alerts
+    self.enterprise_condition = ->(*) do
+      (NotificationSetting.date_alert_settings & notifications.flat_map(&:keys)).any?
```
Can you extract it in a method so that it explains the intent? something like `date_alert_notifications_provided?`

cbliard reviewed (changes requested):
It is working almost as expected. 

With API, all ok: with enterprise token, it's possible to update date alerts notification settings, and without enterprise token, it's not possible

But with GUI, it does not work that well: with enterprise token it works, but without it, when saving user preferences from the user notifications settings page, it displays an error: "is only available in the OpenProject Enterprise Edition".
![image](https://user-images.githubusercontent.com/176055/200329380-e2e7171c-c425-45a7-b4cc-dca370fb6cc0.png)

- the error message should not be displayed as I did not submit anything related to date alerts.
  I can see the paragraph "Date alerts" but there are no checkboxes or drop down.
- the error message is missing a subject.

Also the error message are not very helpful when using the API, as it only states `"Wrong value for Start date, Due date, or Overdue."` without telling if it's a global setting or project setting, or which property, but I think it's still good enough.

dombesz commented:
@cbliard, fixed the issue of displaying the error message while saving the notification settings. Added the feature name into the enterprise error message. Added the EE banner on the settings page.

dombesz committed:
Refactor date alert settings tests

dombesz committed:
Merge branch 'dev' into implementation/44414-add-the-enterprise-edition-guard-on-date-alerts-settings

cbliard reviewed (approved):
Nice :raised_hands:
